<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Overview of tidygenclust • tidygenclust</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Overview of tidygenclust">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidygenclust</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/tidygenclust.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Overview of tidygenclust</h1>
            
      

      <div class="hidden name"><code>tidygenclust.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="using-tidygenclust-for-clustering-in-r">Using <code>tidygenclust</code> for clustering in R<a class="anchor" aria-label="anchor" href="#using-tidygenclust-for-clustering-in-r"></a>
</h2>
<p><code>tidygenclust</code> provides functions and methods to run
genetic clustering in R, using the commonly used ADMIXTURE program, as
well as the python package <code>fastmixture</code>. It also helps to
align and compare multiple runs of the same or different K using the
functionalities of the python package <code>Clumppling</code>. This
package builds on <code>tidypopgen</code>, enhancing the grammar of
population genetics with a focus on genetic clustering.</p>
<p><code>fastmixture</code> (<a href="https://github.com/Rosemeis/fastmixture" class="external-link uri">https://github.com/Rosemeis/fastmixture</a>) (Santander et
al. 2024) is a python tool for ancestry estimation using genotype data.
It uses a model-based approach with comparable accuracy to that of of
popular clustering tool ADMIXTURE, but over considerably faster time
scales making it a more scaleable option for large genetic data
sets.</p>
<p><code>Clumppling</code> (<a href="https://github.com/PopGenClustering/Clumppling" class="external-link uri">https://github.com/PopGenClustering/Clumppling</a>) (Liu,
Kopelman, &amp; Rosenberg, 2024) is a python package which aligns
multiple runs of a clustering analyses allowing for easy comparison and
summary of clustering results both across different values of K and for
repeated runs within K.</p>
<p>With the <code>tidygenclust</code> package we provide a direct R
interface to the python based <code>fastmixture</code> using
<code>reticulate</code> and integrate it with the
<code>tidypopgen</code> package for easy genetic data manipulation
entirely within R. Additionally, the outputs from clustering analyses
(irrespective of the algorithm used) can be easily visualised using the
<code>gt_clumppling</code> function, which allows for easy alignment and
comparison of multiple clustering results.</p>
<p>Genetic data analyses through the integration of
<code>tidypopgen</code> and <code>tidygenclust</code> seamlessly joins
upstream genetic data manipulation (<code>tidypopgen</code>) with
downstream clustering analyses and extensive visualisation options of
the clustering results (<code>tidygenclust</code>) in an integrated
pipeline all natively within R.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>We use <code>reticulate</code> to seamlessly integrate the required
python packages into R, this means that you, as a user, should not need
to worry about the details of the python packages or dependencies.</p>
<p>The first time we run <code>tidygenclust</code>, we need to install
the python packages. We can do this by simply running the following
commands:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">tidygenclust</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tgc_tools_install.html">tgc_tools_install</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="an-example-workflow">An example workflow<a class="anchor" aria-label="anchor" href="#an-example-workflow"></a>
</h2>
<p>To explore the use of <code>tidygenclust</code> with
<code>tidypopgen</code>, we will investigate the genetic ancestry of the
anolis lizard <em>Anolis punctatus</em> across its range in South
America, using data from Prates et al 2018.</p>
<p>We downloaded the vcf file of the genotype data from the Prates et al
2018 GitHub repository which can be accessed <a href="%22%3Chttps://github.com/ivanprates/2018_Anolis_EcolEvol/blob/master/data/VCFtools_SNMF_punctatus_t70_s10_n46/punctatus_t70_s10_n46_filtered.recode.vcf?raw=true%3E%22" class="external-link">here</a>
and compressed it to a vcf.gz file.</p>
<div class="section level3">
<h3 id="read-data-into-gen_tibble-format">Read data into <code>gen_tibble</code> format<a class="anchor" aria-label="anchor" href="#read-data-into-gen_tibble-format"></a>
</h3>
<p>First, using <code>tidypopgen</code> we can read our compressed vcf
file directly into R to create our <code>gen_tibble</code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/EvolEcolGroup/tidypopgen" class="external-link">tidypopgen</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: dplyr</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'dplyr'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     filter, lag</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect, setdiff, setequal, union</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: tibble</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vcf_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="st">"/extdata/anolis/punctatus_t70_s10_n46_filtered.recode.vcf.gz"</span>,</span>
<span>  package <span class="op">=</span> <span class="st">"tidypopgen"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://evolecolgroup.github.io/tidypopgen/reference/gen_tibble.html" class="external-link">gen_tibble</a></span><span class="op">(</span><span class="va">vcf_path</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">TRUE</span>, backingfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="st">"anolis_"</span><span class="op">)</span>,</span>
<span>  parser <span class="op">=</span> <span class="st">"cpp"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>By inspecting our <code>gen_tibble</code> we can see that we have a
total of 46 individuals and 3249 loci, but no population information
attached yet:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A gen_tibble: 3249 loci</span></span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble:     46 × 2</span></span></span>
<span><span class="co">##    id                 genotypes</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> punc_BM288         [0,0,...]</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> punc_GN71          [2,0,...]</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> punc_H1907         [0,2,...]</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> punc_H1911         [0,2,...]</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> punc_H2546         [0,1,...]</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> punc_IBSPCRIB0361  [0,0,...]</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> punc_ICST764       [0,0,...]</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> punc_JFT459        [0,0,...]</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> punc_JFT773        [0,0,...]</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> punc_LG1299        [0,0,...]</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 36 more rows</span></span></span></code></pre>
<p>We can easily attach the population metadata to our
<code>gen_tibble</code> which is stored in anther file that can be found
on the Prates et al 2018 GitHub repository <a href="%22%3Chttps://github.com/ivanprates/2018_Anolis_EcolEvol/raw/master/data/plot_order_punctatus_n46.csv%3E" class="external-link">here</a>.</p>
<p>We can read this file into R and attach the population information to
our <code>gen_tibble</code>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pops_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"/extdata/anolis/punctatus_n46_meta.csv"</span>,</span>
<span>  package <span class="op">=</span> <span class="st">"tidypopgen"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">pops</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="va">pops_path</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">46</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">5</span></span></span>
<span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## <span style="font-weight: bold;">Delimiter:</span> ","</span></span>
<span><span class="co">## <span style="color: #BB0000;">chr</span> (3): id, population, pop</span></span>
<span><span class="co">## <span style="color: #00BB00;">dbl</span> (2): longitude, latitude</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">pops</span>, by <span class="op">=</span> <span class="st">"id"</span><span class="op">)</span></span></code></pre></div>
<p>Now we can inspect the <code>gen_tibble</code> object again to see
that the population information has been added to our genotypes:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A gen_tibble: 3249 loci</span></span></span>
<span><span class="co">## <span style="color: #949494;"># A tibble:     46 × 6</span></span></span>
<span><span class="co">##    id                 genotypes population       longitude latitude pop  </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;vctr_SNP&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> punc_BM288         [0,0,...] Amazonian_Forest     -<span style="color: #BB0000;">51.8</span>    -<span style="color: #BB0000;">3.32</span> Eam  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> punc_GN71          [2,0,...] Amazonian_Forest     -<span style="color: #BB0000;">54.6</span>    -<span style="color: #BB0000;">9.73</span> Eam  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> punc_H1907         [0,2,...] Amazonian_Forest     -<span style="color: #BB0000;">64.8</span>    -<span style="color: #BB0000;">9.45</span> Wam  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> punc_H1911         [0,2,...] Amazonian_Forest     -<span style="color: #BB0000;">64.8</span>    -<span style="color: #BB0000;">9.44</span> Wam  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> punc_H2546         [0,1,...] Amazonian_Forest     -<span style="color: #BB0000;">65.4</span>    -<span style="color: #BB0000;">9.60</span> Wam  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> punc_IBSPCRIB0361  [0,0,...] Atlantic_Forest      -<span style="color: #BB0000;">46.0</span>   -<span style="color: #BB0000;">23.8</span>  AF   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> punc_ICST764       [0,0,...] Atlantic_Forest      -<span style="color: #BB0000;">36.3</span>    -<span style="color: #BB0000;">9.81</span> AF   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> punc_JFT459        [0,0,...] Atlantic_Forest      -<span style="color: #BB0000;">40.5</span>   -<span style="color: #BB0000;">20.3</span>  AF   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> punc_JFT773        [0,0,...] Atlantic_Forest      -<span style="color: #BB0000;">40.5</span>   -<span style="color: #BB0000;">20.0</span>  AF   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> punc_LG1299        [0,0,...] Atlantic_Forest      -<span style="color: #BB0000;">39.1</span>   -<span style="color: #BB0000;">15.3</span>  AF   </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 36 more rows</span></span></span></code></pre>
<p>Finally, we group our <code>gen_tibble</code> by population to make
it easier to plot later, as the grouping information will be passed to
objects created by clustering algorithms:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">population</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="data-preparation-and-pca">Data preparation and PCA<a class="anchor" aria-label="anchor" href="#data-preparation-and-pca"></a>
</h3>
<p>To get an initial idea of our data and potentially help choose a
reasonable starting value for K we may want to run a principal component
analyses (PCA) to explore the data prior to running the
<code>fastmixture</code> analyses.</p>
<p>Before running the PCA we also need to impute any missing values that
may be present in our data.</p>
<p>We can quickly and easily impute and perform a PCA on our
<code>gen_tibble</code> with <code>tidypopgen</code>:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt</span> <span class="op">&lt;-</span> <span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://evolecolgroup.github.io/tidypopgen/reference/gt_impute_simple.html" class="external-link">gt_impute_simple</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"mode"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">anole_pca</span> <span class="op">&lt;-</span> <span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://evolecolgroup.github.io/tidypopgen/reference/gt_pca_partialSVD.html" class="external-link">gt_pca_partialSVD</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">anole_pca</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"scores"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">anole_gt</span><span class="op">$</span><span class="va">population</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"Population"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tidygenclust_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>From the PCA plot we can see that the Amazonian forest and Atlantic
forest populations separate quite clearly. Additionally, whilst the
Atlantic forest individuals cluster tightly together the Amazonian
forest individuals are quite spread indicating three major clusters in
the anolis.</p>
</div>
<div class="section level3">
<h3 id="running-the-fastmixture-algorithm">Running the fastmixture algorithm<a class="anchor" aria-label="anchor" href="#running-the-fastmixture-algorithm"></a>
</h3>
<p>Now we can run <code>gt_fastmixture</code> on our
<code>gen_tibble</code> objec t(note that <code>gt_fastmixture</code>
can also work directly on a PLINK bed file).</p>
<p>As a minimum the <code>gt_fastmixture</code> command requires you to
supply your input data, in this case a <code>gen_tibble</code>, and
specify a value for the number of clusters K.</p>
<p>Based on our PCA results K = 3 may be a good place to start:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_res</span> <span class="op">&lt;-</span> <span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/gt_fastmixture.html">gt_fastmixture</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## -------------------------------------------------</span></span>
<span><span class="co">## fastmixture v1.0.3</span></span>
<span><span class="co">## C.G. Santander, A. Refoyo-Martinez and J. Meisner</span></span>
<span><span class="co">## K=3, seed=42, batches=32, threads=1</span></span>
<span><span class="co">## -------------------------------------------------</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Loaded 46 samples and 3249 SNPs.</span></span>
<span><span class="co">## Random initialization.</span></span>
<span><span class="co">## Initial log-like: -220422.9</span></span>
<span><span class="co">## Performed priming iteration  (0.0s)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Estimating Q and P using mini-batch EM.</span></span>
<span><span class="co">## Using 32 mini-batches.</span></span>
<span><span class="co">## (5)  Log-like: -58511.2  (0.0s)</span></span>
<span><span class="co">## (10) Log-like: -58499.3  (0.0s)</span></span>
<span><span class="co">## (15) Log-like: -58487.9  (0.0s)</span></span>
<span><span class="co">## (20) Log-like: -58483.5  (0.0s)</span></span>
<span><span class="co">## (25) Log-like: -58472.6  (0.0s)</span></span>
<span><span class="co">## (30) Log-like: -58453.3  (0.0s)</span></span>
<span><span class="co">## (35) Log-like: -58458.6  (0.0s)</span></span>
<span><span class="co">## Halving mini-batches to 16.</span></span>
<span><span class="co">## (40) Log-like: -58447.8  (0.0s)</span></span>
<span><span class="co">## (45) Log-like: -58444.6  (0.0s)</span></span>
<span><span class="co">## (50) Log-like: -58465.7  (0.0s)</span></span>
<span><span class="co">## Turning on safety updates.</span></span>
<span><span class="co">## (55) Log-like: -58442.5  (0.0s)</span></span>
<span><span class="co">## (60) Log-like: -58444.4  (0.0s)</span></span>
<span><span class="co">## Halving mini-batches to 8.</span></span>
<span><span class="co">## (65) Log-like: -58439.3  (0.0s)</span></span>
<span><span class="co">## (70) Log-like: -58438.9  (0.0s)</span></span>
<span><span class="co">## Halving mini-batches to 4.</span></span>
<span><span class="co">## (75) Log-like: -58438.0  (0.0s)</span></span>
<span><span class="co">## (80) Log-like: -58437.8  (0.0s)</span></span>
<span><span class="co">## Halving mini-batches to 2.</span></span>
<span><span class="co">## (85) Log-like: -58437.7  (0.0s)</span></span>
<span><span class="co">## (90) Log-like: -58437.7  (0.0s)</span></span>
<span><span class="co">## Running standard updates.</span></span>
<span><span class="co">## No improvement. Returning with best estimate!</span></span>
<span><span class="co">## Final log-likelihood: -58437.6</span></span>
<span><span class="co">## Total elapsed time: 0m0s</span></span></code></pre>
<p>This will <em>very</em> quickly return a single Q matrix, for one
specified K value. But, most likely we want to explore multiple values
for K and should run multiple repeats of each K to assess the stability
of the clustering. The <code>gt_fastmixture</code> function allows you
to specify a vector of K values you wish to run and the number of
repeats per K value.</p>
<p>Now we are doing multiple repeat runs it is also important that for
each repeat we specify a different seed number to ensure consistent and
robust results.</p>
<p>Let’s now set values of K from 2 to 4 and run 3 repeats of each K
value, setting a different random seed for each repeat. We also set the
option <code>no_freqs</code> to <code>FALSE</code> to include
P-matrices, containing ancestral allele frequencies, to our output:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_res</span> <span class="op">&lt;-</span> <span class="va">anole_gt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/gt_fastmixture.html">gt_fastmixture</a></span><span class="op">(</span></span>
<span>  k <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span>, n_runs <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  seed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">42</span>, <span class="fl">2</span>, <span class="fl">16</span><span class="op">)</span>, no_freqs <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Our results are returned as a <code>gt_admix</code> object which
neatly packages the outputted Q matrices, and the corresponding K value
for that run in a structured list.</p>
<p>We can get a summary of our <code>gt_admix</code> results object to
see exactly what it contains:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Admixture results for multiple runs:       </span></span>
<span><span class="co">## k 2 3 4</span></span>
<span><span class="co">## n 3 3 3</span></span>
<span><span class="co">## with slots:</span></span>
<span><span class="co">## $Q for Q matrices</span></span>
<span><span class="co">## $P for  matrices</span></span></code></pre>
<p>From the summary we can see our <code>gt_admix</code> object contains
Q and P matrices for 3 repeat runs of K values 2, 3 and 4 as
expected.</p>
<p>We may want to inspect a specific Q or P matrix in our
<code>gt_admix</code> object and this can be done using the
<code>get_q_matrix</code> or <code>get_p_matrix</code> functions, we
simply need to specify the K value and the repeat run number we are
interested in.</p>
<p>For example we can view the Q matrix corresponding to the second run
of K = 4 like so:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://evolecolgroup.github.io/tidypopgen/reference/get_q_matrix.html" class="external-link">get_q_matrix</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">4</span>, run <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##               .Q1          .Q2          .Q3          .Q4</span></span>
<span><span class="co">## [1,] 9.999923e-06 9.999923e-06 9.999923e-06 9.999700e-01</span></span>
<span><span class="co">## [2,] 9.999914e-06 9.999914e-06 9.999914e-06 9.999700e-01</span></span>
<span><span class="co">## [3,] 3.179351e-01 6.820449e-01 9.999912e-06 9.999912e-06</span></span>
<span><span class="co">## [4,] 9.999700e-01 9.999868e-06 9.999868e-06 9.999868e-06</span></span>
<span><span class="co">## [5,] 2.626097e-01 7.373703e-01 9.999908e-06 9.999908e-06</span></span>
<span><span class="co">## [6,] 9.999845e-06 9.999845e-06 9.999700e-01 9.999845e-06</span></span></code></pre>
<p>Or the P matrix of the first run of K = 3:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://evolecolgroup.github.io/tidypopgen/reference/get_p_matrix.html" class="external-link">get_p_matrix</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">3</span>, run <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           [,1]      [,2]  [,3]</span></span>
<span><span class="co">## [1,] 0.3920996 0.0000100 1e-05</span></span>
<span><span class="co">## [2,] 0.1677997 0.8351086 1e-05</span></span>
<span><span class="co">## [3,] 0.0000100 0.8396586 1e-05</span></span>
<span><span class="co">## [4,] 0.1833421 0.4999566 1e-05</span></span>
<span><span class="co">## [5,] 0.3321308 0.0000100 1e-05</span></span>
<span><span class="co">## [6,] 0.2612136 0.9999900 1e-05</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="visualising-the-results">Visualising the results<a class="anchor" aria-label="anchor" href="#visualising-the-results"></a>
</h3>
<p>For a quick visualisation of a single Q matrix we can use the
autoplot function in <code>tidypopgen</code>:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"barplot"</span>, k <span class="op">=</span> <span class="fl">3</span>, run <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="tidygenclust_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>It is possible to rearrange individual within groups according to
their ancestral components. This makes for a visually appealing plots
that focuses on the main ancestral component within each plot, but makes
multiple plots not comparable (as individuals will not be in the same
order across plots):</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span></span>
<span>  type <span class="op">=</span> <span class="st">"barplot"</span>, k <span class="op">=</span> <span class="fl">3</span>, run <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  reorder_within_groups <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="tidygenclust_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>Note that the colours assigned to each component are arbitrary and
may (and in this case did) change if we reorder individuals. If you want
complete control of your plot, you can create your own customised plot
with <code>ggplot2</code>; we can use <code><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy()</a></code> to easily
extract the required information from a <code>gt_admix</code> object to
use for the plot:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_q_tbl</span> <span class="op">&lt;-</span> <span class="va">anole_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://evolecolgroup.github.io/tidypopgen/reference/get_q_matrix.html" class="external-link">get_q_matrix</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">3</span>, run <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">anole_gt</span><span class="op">)</span></span>
<span><span class="va">anole_q_tbl</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 138 × 4</span></span></span>
<span><span class="co">##    id         group            q     percentage</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> punc_BM288 Amazonian_Forest .Q1   1.000     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> punc_BM288 Amazonian_Forest .Q2   0.000<span style="text-decoration: underline;">010</span>00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> punc_BM288 Amazonian_Forest .Q3   0.000<span style="text-decoration: underline;">010</span>00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> punc_GN71  Amazonian_Forest .Q1   1.000     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> punc_GN71  Amazonian_Forest .Q2   0.000<span style="text-decoration: underline;">010</span>00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> punc_GN71  Amazonian_Forest .Q3   0.000<span style="text-decoration: underline;">010</span>00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> punc_H1907 Amazonian_Forest .Q1   0.000<span style="text-decoration: underline;">010</span>00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> punc_H1907 Amazonian_Forest .Q2   1.000     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> punc_H1907 Amazonian_Forest .Q3   0.000<span style="text-decoration: underline;">010</span>00</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> punc_H1911 Amazonian_Forest .Q1   0.000<span style="text-decoration: underline;">010</span>00</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 128 more rows</span></span></span></code></pre>
<p>In the Prates et al 2018 study, the anolis lizards were found to
split into three genetic groups, two in the Amazonian forest; the
Eastern Amazonia (Eam) and Western Amazonia (Wam) and one in the
Atlantic Forest (AF).</p>
<p>If we wanted to change the grouping variable of our plot to match
these groups we can use the <code>gt_admix_reorder_q</code> function
which will reorder the Q matrix by a chosen grouping variable. Our
metadata contains this new grouping information in the column
<code>pop</code>:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt_admix</span> <span class="op">&lt;-</span> <span class="va">anole_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://evolecolgroup.github.io/tidypopgen/reference/gt_admix_reorder_q.html" class="external-link">gt_admix_reorder_q</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">anole_gt</span><span class="op">$</span><span class="va">pop</span><span class="op">)</span></span></code></pre></div>
<p>We can then visualise our results again with this new grouping:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_gt_admix</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span></span>
<span>  type <span class="op">=</span> <span class="st">"barplot"</span>, k <span class="op">=</span> <span class="fl">3</span>, run <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  reorder_within_groups <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="tidygenclust_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="visualisations-with-gt_clumppling">Visualisations with <code>gt_clumppling</code><a class="anchor" aria-label="anchor" href="#visualisations-with-gt_clumppling"></a>
</h3>
<p>Whilst this gives us a useful, quick visual to check one particular Q
matrix, ideally we want to be able to compare the different K values we
tried in our clustering analyses and to assess the stability of the
multiple repeat runs. For proper comparative visualisation and
assessment of the different K values and repeats we can use the
<code>gt_clummpling</code> function which aligns multiple clustering
results within and between different values of K and allows for easy
visualisation in multipartite plots.</p>
<p>Once we use <code>gt_clumppling</code>, the resulting plots will only
show individuals in the order in which they were found in the
<code>gt_admix</code> object. This means that, if they were not ordered
into groups, we will not be able to annotate groups. One solution would
be to arrange our original gen_tibble by group before starting the
analysis, but we can also use <code>gt_admix_reorder_q</code> to reorder
the individuals in the groups<br>
which we have done in the previous step.</p>
<p>Now let’s run the <code>Clumppling</code> analysis on our
<code>gt_admix</code> object:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_clump</span> <span class="op">&lt;-</span> <span class="va">anole_gt_admix</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/gt_clumppling.html">gt_clumppling</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Now we can visualise the aligned Q matrices for each value of K we
tried:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anole_clump</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"all_modes"</span>, group <span class="op">=</span> <span class="va">anole_gt_admix</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span></code></pre></div>
<p><img src="tidygenclust_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<p>We can see that for the anolis dataset there is only one mode for
each value of K. For a more complex example, where multiple modes are
found in runs of the same K, we can explore a dataset investigating the
the ancestry of Cape Verde individuals, the same example used in the
<code>Clumppling</code> manual.</p>
<p>This time we will read in the Q matrices directly from a directory.
The text files containing the matrices are stored as a zip archive,
which we can pass directly to <code><a href="../reference/gt_clumppling.html">gt_clumppling()</a></code>:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">input_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/capeverde.zip"</span>, package <span class="op">=</span> <span class="st">"tidygenclust"</span><span class="op">)</span></span>
<span><span class="va">clump_res</span> <span class="op">&lt;-</span> <span class="va">input_path</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/gt_clumppling.html">gt_clumppling</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Once we have a <code>gt_clumppling</code> results object, we can use
<code>autoplot</code> to make a number of default plots.</p>
<p>We can plot the modes for all values of k with:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clump_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"all_modes"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tidygenclust_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<p>It is often informative to overlay information on the population from
which each individual was sampled. This can be done by providing a
vector of population labels for each individual. In the case of the Cape
Verde dataset, we have such a vector stored in the package:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">capeverde_pops</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Mandenka" "Mandenka" "Mandenka" "Mandenka" "Mandenka" "Mandenka"</span></span></code></pre>
<p>Let’s get a summary:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">capeverde_pops</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## .</span></span>
<span><span class="co">##      British Cape Verdean       French      Gambian      Iberian     Mandenka </span></span>
<span><span class="co">##           89           44           28          109          107           22</span></span></code></pre>
<p>We can now add it to our plots with the <code>group</code>
argument:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clump_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"all_modes"</span>, group <span class="op">=</span> <span class="va">capeverde_pops</span><span class="op">)</span></span></code></pre></div>
<p><img src="tidygenclust_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<p>And subset our visualisation to only the major modes with:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clump_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"major_modes"</span>, group <span class="op">=</span> <span class="va">capeverde_pops</span><span class="op">)</span></span></code></pre></div>
<p><img src="tidygenclust_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<p>The modes of a specific k value can be plotted with:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clump_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"modes_within_k"</span>, k <span class="op">=</span> <span class="fl">4</span>, group <span class="op">=</span> <span class="va">capeverde_pops</span><span class="op">)</span></span></code></pre></div>
<p><img src="tidygenclust_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
<p>We can also visualise the relationship among modes by plotting over a
multipartite graph, where better alignment between the modes is
indicated by the darker color of the edges connecting their structure
plots (i.e edges with a lower cost of optimal alignment are labelled on
each edge):</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clump_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"modes"</span>, group <span class="op">=</span> <span class="va">capeverde_pops</span><span class="op">)</span></span></code></pre></div>
<p><img src="tidygenclust_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
<p><code>gt_clumppling</code> can also deal with very large K values and
gaps in the K values explored.</p>
<p>For example, we can use the ‘chicken_gapK’ data example also from
<code>Clumppling</code>, which uses outputs from STRUCTURE (so we need
to specify the <code>input_format</code>):</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">input_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/chicken_gapk.zip"</span>, package <span class="op">=</span> <span class="st">"tidygenclust"</span><span class="op">)</span></span>
<span><span class="va">chicken_res</span> <span class="op">&lt;-</span> <span class="va">input_path</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/gt_clumppling.html">gt_clumppling</a></span><span class="op">(</span>input_format <span class="op">=</span> <span class="st">"structure"</span><span class="op">)</span></span></code></pre></div>
<p>We can plot the membership plots on top of the multipartite plot
with:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chicken_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="tidygenclust_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="making-custom-plots-with-ggplot2">Making custom plots with <code>ggplot2</code><a class="anchor" aria-label="anchor" href="#making-custom-plots-with-ggplot2"></a>
</h3>
<p>If we want to customise the default plots, we can use
<code><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy()</a></code> to easily extract the required information from a
<code>gt_clumppling</code> object.</p>
<p>To get all the modes for each k, we use:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clump_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span>matrix <span class="op">=</span> <span class="st">"modes"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 7 × 3</span></span></span>
<span><span class="co">##       k     m label</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span>     2     1 K2M1 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span>     3     1 K3M1 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span>     4     1 K4M1 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span>     4     2 K4M2 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span>     5     1 K5M1 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span>     5     2 K5M2 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">7</span>     5     3 K5M3</span></span></code></pre>
<p>The major modes can be obtained simply with:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clump_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span>matrix <span class="op">=</span> <span class="st">"major_modes"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 4 × 3</span></span></span>
<span><span class="co">##       k     m label</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span>     2     1 K2M1 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span>     3     1 K3M1 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span>     4     1 K4M1 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span>     5     1 K5M1</span></span></code></pre>
<p>To create custom plots, we can also extract the Q matrices for modes,
with their clusters aligned, by typing:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q_modes</span> <span class="op">&lt;-</span> <span class="va">clump_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span>matrix <span class="op">=</span> <span class="st">"Q_modes"</span><span class="op">)</span></span></code></pre></div>
<p><code>q_modes</code> is a list of tibbles, one per mode:</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q_modes</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "K2M1" "K3M1" "K4M1" "K4M2" "K5M1" "K5M2" "K5M3"</span></span></code></pre>
<p>If we only want the major modes, we can simply use:</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q_major_modes</span> <span class="op">&lt;-</span> <span class="va">clump_res</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">tidy</a></span><span class="op">(</span>matrix <span class="op">=</span> <span class="st">"Q_major_modes"</span><span class="op">)</span></span>
<span><span class="va">q_major_modes</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "K2M1" "K3M1" "K4M1" "K5M1"</span></span></code></pre>
<p>Let us inspect one of the tidied modes:</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q4_tidied</span> <span class="op">&lt;-</span> <span class="va">q_major_modes</span><span class="op">[[</span><span class="st">"K4M1"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">q4_tidied</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 3</span></span></span>
<span><span class="co">##      id q     percentage</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span>     1 1      0.000<span style="text-decoration: underline;">010</span>1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span>     1 2      0.000<span style="text-decoration: underline;">010</span>1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span>     1 3      0.000<span style="text-decoration: underline;">010</span>0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span>     1 4      1.000    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span>     2 1      0.057<span style="text-decoration: underline;">4</span>   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span>     2 2      0.000<span style="text-decoration: underline;">010</span>0</span></span></code></pre>
<p>We can now create a simple plot of this mode with:</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="co"># set up the ggplot object</span></span>
<span><span class="va">plt</span> <span class="op">&lt;-</span> <span class="va">q4_tidied</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">id</span>,</span>
<span>    y <span class="op">=</span> <span class="va">percentage</span>,</span>
<span>    fill <span class="op">=</span> <span class="va">q</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># add the columns based on percentage membership to each cluster</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_col</a></span><span class="op">(</span></span>
<span>    width <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    position <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_stack.html" class="external-link">position_stack</a></span><span class="op">(</span>reverse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># set the y label</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"K = 4"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># use a theme to match the distruct look, removing most decorations</span></span>
<span>  <span class="fu"><a href="https://evolecolgroup.github.io/tidypopgen/reference/theme_distruct.html" class="external-link">theme_distruct</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># set the colour scale to be the same as in distruct and clumppling</span></span>
<span>  <span class="fu"><a href="https://evolecolgroup.github.io/tidypopgen/reference/scale_fill_distruct.html" class="external-link">scale_fill_distruct</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plt</span></span></code></pre></div>
<p><img src="tidygenclust_files/figure-html/unnamed-chunk-35-1.png" width="700"></p>
<p>We used a preset theme and colour scale, but you could use any custom
option you prefer by using standard <code>ggplot2</code> theme and
scale_fill options. For example:</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plt</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_d</a></span><span class="op">(</span>guide <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">## Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span></code></pre>
<p><img src="tidygenclust_files/figure-html/unnamed-chunk-36-1.png" width="700"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Eirlys Tysall, Anahit Hovhannisyan, Evie Carter, Andrea Pozzi, Cecilia Padilla-Iglesias, Michela Leonardi, Aramish Fatima, Ondrej Pelanek, Nile Stephenson, Margherita Colucci, Andrea Manica.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
